File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter9/Lesson5/Chapter9_Lesson5.java
import org.ejml.simple.SimpleMatrix;

public class FrameConsistency {

    static SimpleMatrix makeT(SimpleMatrix R, SimpleMatrix p) {
        SimpleMatrix T = SimpleMatrix.identity(4);
        T.insertIntoThis(0,0,R);
        T.insertIntoThis(0,3,p);
        return T;
    }

    static SimpleMatrix invT(SimpleMatrix T) {
        SimpleMatrix R = T.extractMatrix(0,3,0,3);
        SimpleMatrix p = T.extractMatrix(0,3,3,4);
        SimpleMatrix Tinv = SimpleMatrix.identity(4);
        Tinv.insertIntoThis(0,0,R.transpose());
        Tinv.insertIntoThis(0,3,R.transpose().scale(-1).mult(p));
        return Tinv;
    }

    static double frobError(SimpleMatrix E) {
        return E.minus(SimpleMatrix.identity(4)).normF();
    }

    public static void main(String[] args) {
        SimpleMatrix R_WB = SimpleMatrix.identity(3);
        SimpleMatrix p_WB = new SimpleMatrix(3,1,true, new double[]{1,0,0});
        SimpleMatrix T_WB = makeT(R_WB, p_WB);

        SimpleMatrix R_BC = SimpleMatrix.identity(3);
        SimpleMatrix p_BC = new SimpleMatrix(3,1,true, new double[]{0,2,0});
        SimpleMatrix T_BC = makeT(R_BC, p_BC);

        SimpleMatrix T_WC_derived = T_WB.mult(T_BC);
        SimpleMatrix T_WC_meas = T_WC_derived.copy();
        T_WC_meas.set(0,3, T_WC_meas.get(0,3) + 0.05);

        SimpleMatrix E = T_WB.mult(T_BC).mult(invT(T_WC_meas));
        System.out.println("Cycle Frobenius error: " + frobError(E));
    }
}

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter9/Lesson5/Chapter9_Lesson5.cpp
#include <iostream>
#include <Eigen/Dense>

using Eigen::Matrix4d;
using Eigen::Matrix3d;
using Eigen::Vector3d;

Matrix4d makeT(const Matrix3d& R, const Vector3d& p) {
    Matrix4d T = Matrix4d::Identity();
    T.block<3,3>(0,0) = R;
    T.block<3,1>(0,3) = p;
    return T;
}

Matrix4d invT(const Matrix4d& T) {
    Matrix3d R = T.block<3,3>(0,0);
    Vector3d p = T.block<3,1>(0,3);
    Matrix4d Tinv = Matrix4d::Identity();
    Tinv.block<3,3>(0,0) = R.transpose();
    Tinv.block<3,1>(0,3) = -R.transpose() * p;
    return Tinv;
}

double frobError(const Matrix4d& E) {
    return (E - Matrix4d::Identity()).norm(); // Frobenius
}

int main() {
    Matrix3d R_WB = Matrix3d::Identity();
    Vector3d p_WB(1,0,0);
    Matrix4d T_WB = makeT(R_WB, p_WB);

    Matrix3d R_BC = Matrix3d::Identity();
    Vector3d p_BC(0,2,0);
    Matrix4d T_BC = makeT(R_BC, p_BC);

    Matrix4d T_WC_derived = T_WB * T_BC;
    Matrix4d T_WC_meas = T_WC_derived;
    T_WC_meas(0,3) += 0.05;

    Matrix4d E = T_WB * T_BC * invT(T_WC_meas);

    std::cout << "Cycle Frobenius error: " << frobError(E) << std::endl;
    return 0;
}
